File: src/builders/packets/icmp_pkt.rs
use std::net::Ipv4Addr;
use crate::builders::packets::Checksum;
use crate::utils::Mac;



pub(crate) struct IcmpPkt {
    buffer: [u8; 42],
}


impl IcmpPkt {

    pub fn new() -> Self {
        let buffer = Self::build_fixed();
        Self { buffer }
    }



    fn build_fixed() -> [u8; 42] {
        let mut buffer = [0u8; 42];

        // Ethernet header (0 - 14)
        buffer[12..14].copy_from_slice(&0x0800u16.to_be_bytes());

        // IP header (14 - 34)
        buffer[14] = (4 << 4) | 5;
        buffer[15] = 0;
        buffer[16..18].copy_from_slice(&28u16.to_be_bytes());
        buffer[18..20].copy_from_slice(&0x1234u16.to_be_bytes());
        buffer[20..22].copy_from_slice(&0x4000u16.to_be_bytes());
        buffer[22] = 64;
        buffer[23] = 1;
        buffer[24..26].copy_from_slice(&0u16.to_be_bytes());

        // ICMP header (34 - 42)
        buffer[34] = 8;
        buffer[35] = 0;
        buffer[36..38].copy_from_slice(&0u16.to_be_bytes());
        buffer[38..40].copy_from_slice(&0x1234u16.to_be_bytes()); 
        buffer[40..42].copy_from_slice(&1u16.to_be_bytes());

        let cksum = Checksum::icmp_checksum(&buffer[34..42]);
        buffer[36..38].copy_from_slice(&cksum.to_be_bytes());

        buffer
    }



    #[inline]
    fn ether_header(
        &mut self,
        src_mac : Mac,
        dst_mac : Mac
    ) {
        self.buffer[..6].copy_from_slice(dst_mac.bytes());
        self.buffer[6..12].copy_from_slice(src_mac.bytes());
        // 12..14 pre built
    }



    #[inline]
    fn ip_header(
        &mut self,
        src_ip : Ipv4Addr,
        dst_ip : Ipv4Addr
    ) {
        // 14..26 pre built
        self.buffer[26..30].copy_from_slice(&src_ip.octets());
        self.buffer[30..34].copy_from_slice(&dst_ip.octets());

        let cksum = Checksum::ipv4_checksum(&self.buffer[14..34]);
        self.buffer[24..26].copy_from_slice(&cksum.to_be_bytes());
    }



    #[inline]
    pub fn l3_pkt(
        &mut self,
        src_ip : Ipv4Addr,
        dst_ip : Ipv4Addr,
    ) 
      -> &[u8]
    {
        self.ip_header(src_ip, dst_ip);
        &self.buffer[14..42]
    }




    #[inline]
    pub fn l2_pkt(
        &mut self,
        src_mac : Mac,
        src_ip  : Ipv4Addr,
        dst_mac : Mac,
        dst_ip  : Ipv4Addr,
    ) 
      -> &[u8]
    {
        self.ip_header(src_ip, dst_ip);
        self.ether_header(src_mac, dst_mac);

        &self.buffer[..42]
    }

}

File: packet/icmp_pkt.go
package packet

import (
	"encoding/binary"
	"net"
)


type IcmpPkt struct {
	buffer [42]byte
}



func NewIcmpPkt() *IcmpPkt {
	p := &IcmpPkt{}
	p.buildFixed()
	return p
}



func (p *IcmpPkt) buildFixed() {
	// Ethernet header (0-13)
	binary.BigEndian.PutUint16(p.buffer[12:14], 0x0800)
	
	// IP header (14-33)
	p.buffer[14] = (4 << 4) | 5
	p.buffer[15] = 0
	binary.BigEndian.PutUint16(p.buffer[16:18], 28)
	binary.BigEndian.PutUint16(p.buffer[18:20], 0x1234)
	binary.BigEndian.PutUint16(p.buffer[20:22], 0x4000)
	p.buffer[22] = 64
	p.buffer[23] = 1
	binary.BigEndian.PutUint16(p.buffer[24:26], 0)

	// ICMP header (34-41)
	p.buffer[34] = 8
	p.buffer[35] = 0
	binary.BigEndian.PutUint16(p.buffer[36:38], 0)
	binary.BigEndian.PutUint16(p.buffer[38:40], 0x1234)
	binary.BigEndian.PutUint16(p.buffer[40:42], 1)

	ck := IcmpSum(p.buffer[34:42])
	binary.BigEndian.PutUint16(p.buffer[36:38], ck)
}



func (p *IcmpPkt) etherHeader(srcMAC, dstMAC net.HardwareAddr) {
	copy(p.buffer[0:6], dstMAC)
	copy(p.buffer[6:12], srcMAC)
}



func (p *IcmpPkt) ipHeader(srcIP, dstIP net.IP) {
	src := srcIP.To4()
	dst := dstIP.To4()
	if src == nil || dst == nil {
		return
	}
	copy(p.buffer[26:30], src)
	copy(p.buffer[30:34], dst)

	ck := Ipv4Sum(p.buffer[14:34])
	binary.BigEndian.PutUint16(p.buffer[24:26], ck)
}



func (p *IcmpPkt) L3Pkt(srcIP, dstIP net.IP) []byte {
	p.ipHeader(srcIP, dstIP)
	return p.buffer[14:42]
}



func (p *IcmpPkt) L2Pkt(
	srcMAC net.HardwareAddr, 
	srcIP  net.IP, 
	dstMAC net.HardwareAddr, 
	dstIP  net.IP,
) []byte {
	p.ipHeader(srcIP, dstIP)
	p.etherHeader(srcMAC, dstMAC)
	return p.buffer[:]
}